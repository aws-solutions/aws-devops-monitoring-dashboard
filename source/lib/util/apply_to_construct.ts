// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CfnFunction } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { Stack } from 'aws-cdk-lib';
import { CfnPolicy, CfnRole } from 'aws-cdk-lib/aws-iam';
import { addCfnSuppressRules } from '@aws-solutions-constructs/core';
import { NagSuppressions } from 'cdk-nag';

/**
 *
 * @param scope
 * @param id
 */
export function ApplyCfnSuppressRulesToLogRetentionResource(scope: Construct, id: string) {
  // Add rules to suppress to log retention lambda
  const logRetentionLambda = Stack.of(scope).node.tryFindChild(id)?.node.findChild('Resource') as CfnFunction;
  if (logRetentionLambda) {
    addCfnSuppressRules(logRetentionLambda, [
      {
        id: 'W58',
        reason: 'Lambda has the required permission to write CloudWatch Logs through a custom policy.'
      },
      {
        id: 'W89',
        reason: 'There is no need to deploy this Lambda to a VPC.'
      },
      {
        id: 'W92',
        reason: 'There is no need for Reserved Concurrency.'
      }
    ]);
  }

  // Add rules to suppress to log retention lambda policy
  const logRetentionPolicy = Stack.of(scope)
    .node.tryFindChild(id)
    ?.node.tryFindChild('ServiceRole')
    ?.node.findChild('DefaultPolicy')
    .node.findChild('Resource') as CfnPolicy;
  if (logRetentionPolicy) {
    addCfnSuppressRules(logRetentionPolicy, [
      {
        id: 'W12',
        reason:
          'Resource * is required by the Lambda Execution role, so that the Lambda can add ResourcePolicies to all required resources.'
      }
    ]);
    NagSuppressions.addResourceSuppressions(logRetentionPolicy, [
      {
        id: 'AwsSolutions-IAM5',
        reason:
          'Resource * is required by the Lambda Execution role, so that the Lambda can add ResourcePolicies to all required resources.'
      }
    ]);
  }

  // Add cdk-nag suppression to log retention service role
  const logRetentionServiceRole = Stack.of(scope)
    .node.tryFindChild(id)
    ?.node.tryFindChild('ServiceRole')
    ?.node.findChild('Resource') as CfnRole;
  if (logRetentionServiceRole) {
    NagSuppressions.addResourceSuppressions(logRetentionServiceRole, [
      {
        id: 'AwsSolutions-IAM4',
        reason: 'The managed policy is automatically generated by CDK itself to enable log retention.'
      }
    ]);
  }
}
