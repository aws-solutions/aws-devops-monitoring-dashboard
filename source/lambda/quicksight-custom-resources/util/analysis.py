# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from tenacity import retry, retry_if_exception_type, stop_after_attempt

from util.helpers import get_quicksight_client
from util.logging import get_logger
from util.quicksight_resource import QuickSightFailure, QuickSightResource
from util.source_entity import SourceEntity

logger = get_logger(__name__)


class Analysis(QuickSightResource):
    def __init__(
        self, quicksight_application=None, data_sets=None, quicksight_template_arn=None, data_source=None, props=None
    ):
        super().__init__(quicksight_application=quicksight_application, type="analysis", props=props)
        self.use_props(props)

        self.data_sets = data_sets
        self.data_source = data_source
        self.quicksight_template_arn = quicksight_template_arn

        self.config_data = dict()
        self._load_config(self.type, ["main"], self.config_data)
        self.source_entity = SourceEntity(
            data_sets, quicksight_template_arn, self.config_data, source_entity_type="SourceTemplate"
        )

    @retry(retry=retry_if_exception_type(QuickSightFailure), stop=stop_after_attempt(3))
    def create(self):
        logger.info(f"requesting quicksight create_analysis: {self.id}")
        quicksight_client = get_quicksight_client()

        try:
            response = quicksight_client.create_analysis(
                AwsAccountId=self.aws_account_id,
                AnalysisId=self.id,
                Name=self.name,
                Permissions=self._get_permissions(),
                SourceEntity=self._get_source_entity(),
            )
            logger.info(f"finished quicksight create_analysis for id:{self.id}, response: {response}")
        except quicksight_client.exceptions.ResourceExistsException:
            response = quicksight_client.describe_analysis(AwsAccountId=self.aws_account_id, AnalysisId=self.id)
            response = response["Analysis"]
        except quicksight_client.exceptions.InvalidParameterValueException as exc:
            logger.error(str(exc))
            raise QuickSightFailure()

        self.arn = response["Arn"]
        return response

    def delete(self):
        logger.info(f"requesting quicksight delete_analysis id:{self.id}")
        quicksight_client = get_quicksight_client()

        response = quicksight_client.delete_analysis(AwsAccountId=self.aws_account_id, AnalysisId=self.id)
        logger.info(f"finished quicksight delete_analysis for id:{self.id}, response: {response}")
        return response

    def _get_permissions(self):
        # The principal is the owner of the resource and create the resources and is given full actions for the type
        permissions = [
            {
                "Principal": self.principal_arn,
                "Actions": [
                    "quicksight:RestoreAnalysis",
                    "quicksight:UpdateAnalysisPermissions",
                    "quicksight:DeleteAnalysis",
                    "quicksight:QueryAnalysis",
                    "quicksight:DescribeAnalysisPermissions",
                    "quicksight:DescribeAnalysis",
                    "quicksight:UpdateAnalysis"
                ],
            }
        ]
        return permissions

    def _get_source_entity(self):
        return self.source_entity.get_source_entity()
