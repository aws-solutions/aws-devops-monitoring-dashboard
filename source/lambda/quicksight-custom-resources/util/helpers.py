# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import json
from os import environ

import boto3
import botocore.config

from util.logging import get_logger

logger = get_logger(__name__)

# Global boto3 clients to help with initialization and performance
_helpers_service_clients = dict()


class EnvironmentVariableError(Exception):
    pass


def get_service_client(service_name):
    """Get the global service boto3 client"""
    global _helpers_service_clients
    if service_name not in _helpers_service_clients:        
        config = botocore.config.Config(retries=dict(max_attempts=3))

        logger.debug(f"Initializing global boto3 client for {service_name}")
        _helpers_service_clients[service_name] = boto3.client(service_name, config=config, region_name=get_aws_region())
    return _helpers_service_clients[service_name]


def get_quicksight_client():
    """Get the global quicksight boto3 client"""
    return get_service_client("quicksight")


def get_sts_client():
    """Get the global sts boto3 client"""
    return get_service_client("sts")


def get_aws_partition():
    """
    Get the caller's AWS partition by driving it from AWS region
    :return: partition name for the current AWS region (e.g. aws)
    """
    region_name = environ.get("AWS_REGION")
    china_region_name_prefix = "cn"
    us_gov_cloud_region_name_prefix = "us-gov"
    aws_regions_partition = "aws"
    aws_china_regions_partition = "aws-cn"
    aws_us_gov_cloud_regions_partition = "aws-us-gov"

    # China regions
    if region_name.startswith(china_region_name_prefix):
        return aws_china_regions_partition
    # AWS GovCloud(US) Regions
    elif region_name.startswith(us_gov_cloud_region_name_prefix):
        return aws_us_gov_cloud_regions_partition
    else:
        return aws_regions_partition


def get_aws_region():
    """
    Get the caller's AWS region from the environment variable AWS_REGION
    :return: the AWS region name (e.g. us-east-1)
    """
    region = environ.get("AWS_REGION")
    if not region:
        raise EnvironmentVariableError("Missing AWS_REGION environment variable.")

    return region


def get_aws_account_id():
    """
    Get the caller's AWS account ID
    :return: The AWS account ID
    """
    sts_client = get_sts_client()
    identity = sts_client.get_caller_identity()
    return identity.get("Account")
