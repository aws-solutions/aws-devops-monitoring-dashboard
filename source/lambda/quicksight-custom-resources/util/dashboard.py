# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from util.helpers import get_quicksight_client
from util.logging import get_logger
from util.quicksight_resource import QuickSightResource
from util.source_entity import SourceEntity

logger = get_logger(__name__)


class Dashboard(QuickSightResource):
    def __init__(
        self,
        quicksight_application=None,
        data_sets=None,
        quicksight_template_arn=None,
        data_source=None,
        props=None,
    ):
        super().__init__(quicksight_application=quicksight_application, type = "dashboard", props=props)
        self.use_props(props)

        self.data_sets = data_sets
        self.data_source = data_source
        self.quicksight_template_arn = quicksight_template_arn

        self.config_data = dict()
        self._load_config(self.type, ["main"], self.config_data)
        self.source_entity = SourceEntity(
            data_sets, quicksight_template_arn, self.config_data, source_entity_type="SourceTemplate"
        )

    def create(self):
        logger.info(f"requesting quicksight create_dashboard: {self.id}")
        quicksight_client = get_quicksight_client()

        response = quicksight_client.create_dashboard(
            AwsAccountId=self.aws_account_id,
            DashboardId=self.id,
            Name=self.name,
            Permissions=self._get_permissions(),
            SourceEntity=self._get_source_entity(),
            DashboardPublishOptions=self._get_dashboard_publish_options(),
        )
        logger.info(f"finished quicksight create_dashboard for id:{self.id}, response: {response}")

        self.arn = response["Arn"]
        return response

    def delete(self):
        logger.info(f"requesting quicksight delete_dashboard id:{self.id}")
        quicksight_client = get_quicksight_client()

        response = quicksight_client.delete_dashboard(AwsAccountId=self.aws_account_id, DashboardId=self.id)
        logger.info(f"finished quicksight delete_dashboard for id:{self.id}, response: {response}")

        return response

    def _get_dashboard_publish_options(self):
        dashboard_publish_options = {
            "AdHocFilteringOption": {"AvailabilityStatus": "ENABLED"},
            "ExportToCSVOption": {"AvailabilityStatus": "ENABLED"},
            "SheetControlsOption": {"VisibilityState": "EXPANDED"},
        }
        return dashboard_publish_options

    def _get_permissions(self):
        # The principal is the owner of the resource and create the resources and is given full actions for the type
        permissions = [
            {
                "Principal": self.principal_arn,
                "Actions": [
                    "quicksight:DescribeDashboard",
                    "quicksight:ListDashboardVersions",
                    "quicksight:UpdateDashboardPermissions",
                    "quicksight:QueryDashboard",
                    "quicksight:UpdateDashboard",
                    "quicksight:DeleteDashboard",
                    "quicksight:DescribeDashboardPermissions",
                    "quicksight:UpdateDashboardPublishedVersion",
                ],
            }
        ]
        return permissions

    def _get_source_entity(self):
        return self.source_entity.get_source_entity()
