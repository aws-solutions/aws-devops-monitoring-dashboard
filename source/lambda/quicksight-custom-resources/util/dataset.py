# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from tenacity import retry, retry_if_exception_type, stop_after_attempt

from util.helpers import get_quicksight_client
from util.logging import get_logger
from util.quicksight_resource import QuickSightFailure, QuickSightResource

logger = get_logger(__name__)


class DataSet(QuickSightResource):
    def __init__(
        self, quicksight_application=None, data_source=None, data_set_sub_type=None, props=None
    ):
        super().__init__(quicksight_application=quicksight_application, type = "dataset", sub_type=data_set_sub_type, props=props)
        self.use_props(props)

        self.data_source = data_source

        # override to set schema name directly
        # instead of using the one in the config file
        self.schema = None

        self.config_data = dict()
        self._load_config(self.type, quicksight_application.get_supported_data_set_sub_types(), self.config_data)

    def create(self):
        if not self.data_source:
            raise ValueError("missing datasource value when creating dataset")
        logger.info(f"creating quicksight dataset id:{self.id}")
        physical_table_map = self._get_map(self.sub_type, "PhysicalTableMap")
        logical_table_map = self._get_map(self.sub_type, "LogicalTableMap")
        response = self._create_data_set(physical_table_map, logical_table_map)
        return response

    def delete(self):
        logger.info(f"deleting quicksight dataset id:{self.id}")
        quicksight_client = get_quicksight_client()

        response = quicksight_client.delete_data_set(AwsAccountId=self.aws_account_id, DataSetId=self.id)
        logger.info(f"finished deleting quicksight dataset for id:{self.id}, response:{response}")

        self.arn = response["Arn"]
        return response

    @retry(retry=retry_if_exception_type(QuickSightFailure), stop=stop_after_attempt(3))
    def _create_data_set(self, physical_table_map, logical_table_map):
        quicksight_client = get_quicksight_client()

        self._update_data_source_arn(physical_table_map)
        self._update_schema(physical_table_map)

        try:
            response = quicksight_client.create_data_set(
                AwsAccountId=self.aws_account_id,
                DataSetId=self.id,
                Name=self.name,
                Permissions=self._get_permissions(),
                PhysicalTableMap=physical_table_map,
                LogicalTableMap=logical_table_map,
                ImportMode="DIRECT_QUERY",
            )
            logger.info(f"finished creating quicksight create_data_set id:{self.id}, response:{response}")
        except quicksight_client.exceptions.ResourceExistsException:
            logger.info(f"dataset for id:{self.id} already exists")
            response = quicksight_client.describe_data_set(AwsAccountId=self.aws_account_id, DataSetId=self.id)
            response = response["DataSet"]
        except quicksight_client.exceptions.InvalidParameterValueException as exc:
            logger.error(str(exc))
            raise QuickSightFailure()

        self.arn = response["Arn"]
        return response

    def _get_permissions(self):
        # The principal is the owner of the resource and create the resources and is given full actions for the type
        permissions = [
            {
                "Principal": self.principal_arn,
                "Actions": [
                    "quicksight:DescribeDataSet",
                    "quicksight:DescribeDataSetPermissions",
                    "quicksight:PassDataSet",
                    "quicksight:DescribeIngestion",
                    "quicksight:ListIngestions",
                    "quicksight:UpdateDataSet",
                    "quicksight:DeleteDataSet",
                    "quicksight:CreateIngestion",
                    "quicksight:CancelIngestion",
                    "quicksight:UpdateDataSetPermissions",
                ],
            }
        ]
        return permissions


    def _update_schema(self, obj):
        if not self.schema:
            logger.debug(f"Schema name is not set in object. Using the ones from config file as is in RelationalTable[].Schema in PhysicalTableMap")
            return
        for (key, value) in obj.items():
            logger.debug(f"Updating schema arn value of RelationalTable.Schema in {key} PhysicalTableMap")
            value["RelationalTable"]["Schema"] = self.schema

    def _update_data_source_arn(self, obj):
        for (key, value) in obj.items():
            logger.debug(f"Updating datasource arn value of RelationalTable.DataSourceArn in {key} PhysicalTableMap")
            value["RelationalTable"]["DataSourceArn"] = self.data_source.arn
