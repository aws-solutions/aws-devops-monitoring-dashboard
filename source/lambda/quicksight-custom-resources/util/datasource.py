# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from util.helpers import get_quicksight_client
from util.logging import get_logger
from util.quicksight_resource import QuickSightResource

logger = get_logger(__name__)


class DataSource(QuickSightResource):
    def __init__(self, quicksight_application=None, props=None):
        super().__init__(quicksight_application, type="datasource", props=props)
        self.use_props(props)
        self.athena_workgroup = "primary"

    def create(self):
        logger.info(f"creating quicksight datasource id:{self.id}")
        quicksight_client = get_quicksight_client()

        data_source_parameters = {"AthenaParameters": {"WorkGroup": self.athena_workgroup}}

        try:
            response = quicksight_client.create_data_source(
                AwsAccountId=self.aws_account_id,
                DataSourceId=self.id,
                Name=self.name,
                Type="ATHENA",
                DataSourceParameters=data_source_parameters,
                Permissions=self._get_permissions(),
                SslProperties={"DisableSsl": False},
            )
            logger.info(f"finished creating quicksight datasource for id:{self.id}, response {response}")
        except quicksight_client.exceptions.ResourceExistsException:
            logger.info(f"datasource for id:{self.id} already exists")
            response = quicksight_client.describe_data_source(AwsAccountId=self.aws_account_id, DataSourceId=self.id)
            response = response["DataSource"]

        self.arn = response["Arn"]
        return response

    def update(self):
        quicksight_client = get_quicksight_client()
        quicksight_client.describe_data_source
        data_source_parameters = {"AthenaParameters": {"WorkGroup": self.athena_workgroup}}
        try:
            response = quicksight_client.update_data_source(
                AwsAccountId=self.aws_account_id,
                DataSourceId=self.id,
                Name=self.name,
                DataSourceParameters=data_source_parameters,
                SslProperties={"DisableSsl": False},
            )
        except quicksight_client.exceptions.ConflictException as exc:
            logger.debug(str(exc))
            response = response["DataSource"]
        return response

    def delete(self):
        logger.info(f"deleting quicksight datasource id:{self.id}")
        quicksight_client = get_quicksight_client()

        response = quicksight_client.delete_data_source(AwsAccountId=self.aws_account_id, DataSourceId=self.id)
        logger.info(f"finished deleting quicksight datasource for id:{self.id}, response:{response}")
        self.arn = response["Arn"]
        return response

    def _get_permissions(self):
        # The principal is the owner of the resource and create the resources and is given full actions for the type
        permissions = [
            {
                "Principal": self.principal_arn,
                "Actions": [
                    "quicksight:DescribeDataSource",
                    "quicksight:DescribeDataSourcePermissions",
                    "quicksight:PassDataSource",
                    "quicksight:UpdateDataSource",
                    "quicksight:UpdateDataSourcePermissions",
                    "quicksight:DeleteDataSource",
                ],
            }
        ]
        return permissions
