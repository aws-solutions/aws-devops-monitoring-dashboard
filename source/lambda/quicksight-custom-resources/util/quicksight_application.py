# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import json

import yaml

from util.analysis import Analysis
from util.dashboard import Dashboard
from util.dataset import DataSet
from util.datasource import DataSource
from util.helpers import get_aws_account_id, get_quicksight_client
from util.logging import get_logger
from util.template import Template

logger = get_logger(__name__)

# Global state. Keep in execution context of lambda
_global_state = dict()


def get_global_state():
    """Get the global state"""
    global _global_state
    if not _global_state:
        logger.debug(f"Initializing global state for quicksight api")
        _global_state = dict()
    return _global_state


def read_config(file_name):
    config_path = file_name
    with open(config_path, "r") as f:
        body = f.read()
    config_dict = yaml.safe_load(body)
    logger.info(f"config loaded, config: {config_dict}")
    return config_dict


class QuicksightApplication:
    def __init__(self, resource_properties):

        supported_data_set_types = ["code-change-activity", "code-deployment-detail", "recovery-time-detail", "code-pipeline-detail", "code-build-detail", "github-change-activity"]

        self.resource_properties = resource_properties
        self.global_state = get_global_state()

        # use config data file if provided
        config_file = resource_properties.get("ConfigDataFile", None)
        if config_file:
            data = read_config(config_file)
            self.global_state.update(data)

        self.prefix = resource_properties.get("StackName", "Sample_Sol")

        self.quicksight_template_arn = resource_properties.get(
            "QuickSightSourceTemplateArn", "Uninitialized QuickSightSourceTemplateArn"
        )
        logger.debug(f"Using QuickSightSourceTemplateArn: {self.quicksight_template_arn }")

        self.quicksight_principal_arn = resource_properties.get(
            "QuickSightPrincipalArn", "Uninitialized QuickSightPrincipalArn"
        )
        logger.debug(f"Using QuickSightPrincipalArn: {self.quicksight_principal_arn }")

        self.data_source = DataSource(quicksight_application=self, props=self.global_state)
        self.data_source.athena_workgroup = resource_properties.get("WorkGroupName", "primary")

        self.data_set_sub_types = supported_data_set_types
        self.data_sets = dict()
        for data_set_sub_type in self.data_set_sub_types:
            data_set = DataSet(
                quicksight_application=self,
                data_source=self.data_source,
                data_set_sub_type=data_set_sub_type,
                props=self.global_state,
            )
            self.data_sets[data_set_sub_type] = data_set

        self.analysis = Analysis(
            quicksight_application=self,
            data_sets=self.data_sets,
            quicksight_template_arn=self.quicksight_template_arn,
            data_source=self.data_source,
            props=self.global_state,
        )

        self.dashboard = Dashboard(
            quicksight_application=self,
            data_source=self.data_source,
            data_sets=self.data_sets,
            quicksight_template_arn=self.quicksight_template_arn,
            props=self.global_state,
        )

        self.template = Template(quicksight_application=self, data_sets=self.data_sets, props=self.global_state)

        global_state_json = json.dumps(self.global_state, indent=2, sort_keys=True)
        logger.debug(f"QuicksightApi: after init, global data json: {global_state_json}")

    def get_data_source(self):
        return self.data_source

    def get_data_sets(self):
        return self.data_sets

    def get_analysis(self):
        return self.analysis

    def get_dashboard(self):
        return self.dashboard

    def get_template(self):
        return self.template

    def get_supported_data_set_sub_types(self):
        return self.data_set_sub_types

    def get_global_state(self):
        return self.global_state

    @property
    def edition(self) -> str:
        qs = get_quicksight_client()
        try:
            settings = qs.describe_account_settings(AwsAccountId=get_aws_account_id())
            edition = settings.get("AccountSettings").get("Edition")
        except qs.exceptions.ResourceNotFoundException:
            edition = "DISABLED"

        logger.info("running with QuickSight %s" % edition)
        return edition

    @staticmethod
    def clear_global_states():
        global _global_state
        _global_state = dict()
