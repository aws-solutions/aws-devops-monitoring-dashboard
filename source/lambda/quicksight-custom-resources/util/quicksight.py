# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from util.logging import get_logger
from util.quicksight_application import QuicksightApplication
from util.template import TemplatePermissionType

logger = get_logger(__name__)


class QuicksightApi:
    def __init__(self, resource_properties):
        self.quicksight_application = QuicksightApplication(resource_properties)
        self.global_state = self.quicksight_application.get_global_state()

    def create_all_resources(self):
        responses = []

        responses.append(self.create_data_source())
        responses.append(self.create_data_sets())
        responses.append(self.create_analysis())
        responses.append(self.create_dashboard())

        return responses

    def create_data_source(self):
        qs_resource = self.quicksight_application.get_data_source()
        response = qs_resource.create()
        self.get_global_state().update({"datasource": qs_resource.get_data()})
        return response

    def create_data_sets(self):
        responses = []

        data_set_sub_types = self.quicksight_application.get_supported_data_set_sub_types()
        data_sets = self.quicksight_application.get_data_sets()
        self.get_global_state().update({"dataset": {}})
        for data_set_type in data_set_sub_types:
            response = data_sets[data_set_type].create()
            responses.append(response)
            self.get_global_state()["dataset"].update({data_set_type: data_sets[data_set_type].get_data()})

        return responses

    def create_analysis(self):
        qs_resource = self.quicksight_application.get_analysis()
        response = qs_resource.create()
        self.get_global_state().update({"analysis": qs_resource.get_data()})
        return response

    def create_dashboard(self):
        qs_resource = self.quicksight_application.get_dashboard()
        response = qs_resource.create()
        self.get_global_state().update({"dashboard": qs_resource.get_data()})
        return response

    def delete_all_resources(self):
        responses = []
        """
        To ensure deletion is done on a best effort basis, each method call is in its own try except block
        Any exception that occurs when deleting is logged as a warning but not raised as an exception to
        continue deleting the other QuickSight resources
        """
        try:
            responses.append(self.delete_dashboard())
        except Exception as error:
            logger.warning(f"Failed to delete dashboard {error}")

        try:
            responses.append(self.delete_analysis())
        except Exception as error:
            logger.warning(f"Failed to delete analysis {error}")

        try:
            responses.append(self.delete_data_sets())
        except Exception as error:
            logger.warning(f"Failed to delete data sets {error}")

        try:
            responses.append(self.delete_data_source())
        except Exception as error:
            logger.warning(f"Failed to delete data source {error}")

        return responses

    def delete_data_source(self):
        qs_resource = self.quicksight_application.get_data_source()
        response = qs_resource.delete()
        return response

    def delete_data_sets(self):
        responses = []

        data_sets = self.quicksight_application.get_data_sets()
        for data_set in data_sets.values():
            response = data_set.delete()
            responses.append(response)
        return responses

    def delete_analysis(self):
        qs_resource = self.quicksight_application.get_analysis()
        response = qs_resource.delete()
        return response

    def delete_dashboard(self):
        qs_resource = self.quicksight_application.get_dashboard()
        response = qs_resource.delete()
        return response

    def create_template_from_template(self, source_template_arn):
        qs_resource = self.quicksight_application.get_template()
        response = qs_resource.create_from_template(source_template_arn)
        self.get_global_state().update({"template": qs_resource.get_data()})
        return response

    def create_template_from_analysis(self):
        template = self.quicksight_application.get_template()
        analysis = self.quicksight_application.get_analysis()
        response = template.create_from_analysis(analysis)
        self.get_global_state().update({"template": template.get_data()})
        return response

    def create_template_from_dashboard(self):
        template = self.quicksight_application.get_template()
        dashboard = self.quicksight_application.get_dashboard()
        response = template.create_from_dashboard(dashboard)
        self.get_global_state().update({"template": template.get_data()})
        return response

    def update_template_permissions(
        self, permission: TemplatePermissionType = TemplatePermissionType.PUBLIC, principal=None
    ):
        qs_resource = self.quicksight_application.get_template()
        response = qs_resource.update_template_permissions(permission, principal)
        return response

    def delete_template(self):
        qs_resource = self.quicksight_application.get_template()
        response = qs_resource.delete()
        return response

    def get_global_state(self):
        return self.global_state

    def describe_data_source(self):
        qs_resource = self.quicksight_application.get_data_source()
        response = qs_resource.describe()
        return response

    def describe_data_sets(self):
        responses = []

        data_sets = self.quicksight_application.get_data_sets()
        for data_set in data_sets.values():
            response = data_set.describe()
            responses.append(response)
        return responses

    def describe_analysis(self):
        qs_resource = self.quicksight_application.get_analysis()
        response = qs_resource.describe()
        return response

    def describe_dashboard(self):
        qs_resource = self.quicksight_application.get_dashboard()
        response = qs_resource.describe()
        return response
