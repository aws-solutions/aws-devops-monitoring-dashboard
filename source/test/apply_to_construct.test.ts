// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ResourcePart, SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cdk from 'aws-cdk-lib';
import { Role, ServicePrincipal, PolicyStatement, Effect, PolicyDocument, CfnRole } from 'aws-cdk-lib/aws-iam';
import { ApplyCfnSuppressRulesToLogRetentionResource } from '../lib/util/apply_to_construct';

test('adding cfn suppressing rules to lambda', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app);

  const cwLogsPS = new PolicyStatement({
    actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
    effect: Effect.ALLOW,
    resources: [
      stack.formatArn({
        service: 'logs',
        resource: 'log-group',
        arnFormat: cdk.ArnFormat.COLON_RESOURCE_NAME,
        resourceName: '/aws/lambda/*'
      })
    ],
    sid: 'CreateCWLogs'
  });

  const lambdaRole = new Role(stack, 'ServiceRole', {
    assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
    path: '/',
    inlinePolicies: {
      LambdaPolicyName: new PolicyDocument({
        statements: [cwLogsPS]
      })
    }
  });

  const cfnLambdaRole = lambdaRole.node.defaultChild as CfnRole;

  cfnLambdaRole.overrideLogicalId('ServiceRole');

  new lambda.Function(stack, 'testFunction', {
    code: lambda.Code.fromInline('exports.handler = (event, context, callback) => {}'),
    runtime: lambda.Runtime.NODEJS_14_X,
    handler: 'index.handler',
    role: lambdaRole
  });

  ApplyCfnSuppressRulesToLogRetentionResource(stack, 'testFunction');

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

  expect(stack).toHaveResourceLike(
    'AWS::Lambda::Function',
    {
      Metadata: {
        cfn_nag: {
          rules_to_suppress: [
            {
              id: 'W58',
              reason: 'Lambda has the required permission to write CloudWatch Logs through a custom policy.'
            },
            {
              id: 'W89',
              reason: 'There is no need to deploy this Lambda to a VPC.'
            },
            {
              id: 'W92',
              reason: 'There is no need for Reserved Concurrency.'
            }
          ]
        }
      }
    },
    ResourcePart.CompleteDefinition
  );
});
