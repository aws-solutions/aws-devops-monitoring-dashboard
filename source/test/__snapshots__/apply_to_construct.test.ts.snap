// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding cfn suppressing rules to lambda 1`] = `
{
  "Resources": {
    "ServiceRole": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                  "Sid": "CreateCWLogs",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaPolicyName",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testFunction483F4CBE": {
      "DependsOn": [
        "ServiceRole",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "Lambda has the required permission to write CloudWatch Logs through a custom policy.",
            },
            {
              "id": "W89",
              "reason": "There is no need to deploy this Lambda to a VPC.",
            },
            {
              "id": "W92",
              "reason": "There is no need for Reserved Concurrency.",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = (event, context, callback) => {}",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ServiceRole",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
  },
}
`;
